db/memtable.cc

```diff
@@ -1672,4 +1672,29 @@ uint64_t MemTable::GetMinLogContainingPrepSection() {
   return min_prep_log_referenced_.load();
 }
 
+const Slice& MemTable::GetNewestUDT() {
+  const Comparator* ucmp = GetInternalKeyComparator().user_comparator();
+  size_t ts_sz = ucmp->timestamp_size();
+  // This path should not be invoked for MemTables that does not enable the UDT
+  // feature.
+  assert(ts_sz > 0);
+  if (!newest_udt_.empty()) {
+    return newest_udt_;
+  }
+
+  Arena arena;
+  ReadOptions ro;
+  InternalIterator* iter = NewIterator(ro, &arena);
+
+  Slice udt;
+  iter->SeekToFirst();
+  for (; iter->Valid(); iter->Next()) {
+    udt = ExtractTimestampFromKey(iter->key(), ts_sz);
+    if (newest_udt_.empty() || ucmp->CompareTimestamp(udt, newest_udt_) > 0) {
+      newest_udt_ = udt;
+    }
+  }
```

Comment:
In order to not affect live writes, this field is only calculated for an immutable Memtable by a background job. Since we need to check if all the user-defined timestamps expired, this need to scan the whole table to get the newest UDT.  I haven't thought of an alternative to this.

Reply:
Thank you for this perspective! I was trying to avoid having regressions on the live write path for users that are not use this feature. But didn't realize that scan table while holding db mutex could have a bigger impact. I have updated the tracking to happen on the write path, for non users, the impact of this branch should be small.